### PR TITLE
AV-89547: go SDK crashing when controller status check polling finishes up unsuccessfully.

### PR DESCRIPTION
**Pull request by mayank-avinetworks**
The SDK HTTP client side error handling needs to be revisited, it is expected that response of HTTP API will always have the body but in case of failures like controllerStatusCheck, where on encountering the client side errors the flow is routed to for recovery, itself is not able to reach server at all, the body will be missing. 

 

Day 1 issue, but got caught after we moved away from infinite poll for controller status check .

This is the minimal code change I figured out to fix the crash, we can take up the revamp of error handling separately in the long run.

=== RUN TestAviSession
--- PASS: TestAviSession (1.67s)
=== RUN TestAviSessionControllerStatusCheckLimits
--- PASS: TestAviSessionControllerStatusCheckLimits (0.00s)
=== RUN TestAviPool
--- PASS: TestAviPool (1.61s)
=== RUN TestTenantSwitch
--- PASS: TestTenantSwitch (2.72s)
=== RUN TestApiLogout
--- PASS: TestApiLogout (4.63s)
=== RUN TestApiReLogin
--- PASS: TestApiReLogin (4.78s)
=== RUN TestAviDefaultFields
--- PASS: TestAviDefaultFields (1.32s)
=== RUN TestTokenAuthRobustness
--- SKIP: TestTokenAuthRobustness (0.00s)
avisession_test.go:404: SKIPPING as test not running in controller.
=== RUN TestTokenAuthRobustnessV2
--- PASS: TestTokenAuthRobustnessV2 (0.03s)
=== RUN TestAviReads
--- PASS: TestAviReads (1.15s)
=== RUN TestApiLazyAuthentication
--- PASS: TestApiLazyAuthentication (0.31s)
PASS
ok github.com/avinetworks/sdk/go/session 18.216s
mayank-dev:/home/aviuser/project/go_sdk/src/github.com/avinetworks/sdk/go/session$

### CODE REVIEW

**Comment by mayank-avinetworks** [comment]
=== RUN   TestAviSession
--- PASS: TestAviSession (1.67s)
=== RUN   TestAviSessionControllerStatusCheckLimits
--- PASS: TestAviSessionControllerStatusCheckLimits (0.00s)
=== RUN   TestAviPool
--- PASS: TestAviPool (1.61s)
=== RUN   TestTenantSwitch
--- PASS: TestTenantSwitch (2.72s)
=== RUN   TestApiLogout
--- PASS: TestApiLogout (4.63s)
=== RUN   TestApiReLogin
--- PASS: TestApiReLogin (4.78s)
=== RUN   TestAviDefaultFields
--- PASS: TestAviDefaultFields (1.32s)
=== RUN   TestTokenAuthRobustness
--- SKIP: TestTokenAuthRobustness (0.00s)
avisession_test.go:404: SKIPPING as test not running in controller.
=== RUN   TestTokenAuthRobustnessV2
--- PASS: TestTokenAuthRobustnessV2 (0.03s)
=== RUN   TestAviReads
--- PASS: TestAviReads (1.15s)
=== RUN   TestApiLazyAuthentication
--- PASS: TestApiLazyAuthentication (0.31s)
PASS
ok  	github.com/avinetworks/sdk/go/session	18.216s
mayank-dev:/home/aviuser/project/go_sdk/src/github.com/avinetworks/sdk/go/session$

**Comment by mayank-avinetworks** [comment]
Requested @sudswasavi  to test it once for his use cases to check if any crash is observed.

File: go/session/avisession.go [hunk 1]
```diff
@@ -553,28 +553,38 @@ func (avisess *AviSession) restRequest(verb string, uri string, payload interfac
...
 		return result, nil
 	}
+	// It cannot be assumed that the error will always be from server side in response.
+	// Error could be from HTTP client side which will not have body in response.
+	// Need to change our API resp handling design if we want to handle client side errors separately.
+
+	// Below block will take care for errors without body.
+	if resp.Body == nil {
+		glog.Errorf("Encountered client side error: %+v", resp)
+		errorResult.Message = &resp.Status
```
Comment (manojkumarjain) [inline go/session/avisession.go:583]: What will be status on client  failure.
Will it have any StatusCode

Comment (manojkumarjain) [reply go/session/avisession.go:583]: Ok found

**Review by manojkumarjain — APPROVED** [review]
Approved

**Comment by mayank-avinetworks** [comment]
mayank-dev:/home/aviuser/project/go_sdk/src/github.com/avinetworks/sdk/go/session$go test -v -run TestAviSessionControllerStatusCheckLimits
=== RUN   TestAviSessionControllerStatusCheckLimits
E0708 04:11:59.380304   10456 avisession.go:860] CheckControllerStatus Error while generating http request https://localhost//api/cluster/status Get https://localhost//api/cluster/status: dial tcp 127.0.0.1:443: connect: connection refused
E0708 04:12:09.380836   10456 avisession.go:870] CheckControllerStatus Controller https://localhost//api/cluster/status Retrying. round 0..!
E0708 04:12:09.388873   10456 avisession.go:860] CheckControllerStatus Error while generating http request https://localhost//api/cluster/status Get https://localhost//api/cluster/status: dial tcp 127.0.0.1:443: connect: connection refused
E0708 04:12:19.389149   10456 avisession.go:870] CheckControllerStatus Controller https://localhost//api/cluster/status Retrying. round 1..!
E0708 04:12:19.396124   10456 avisession.go:860] CheckControllerStatus Error while generating http request https://localhost//api/cluster/status Get https://localhost//api/cluster/status: dial tcp 127.0.0.1:443: connect: connection refused
E0708 04:12:29.396508   10456 avisession.go:870] CheckControllerStatus Controller https://localhost//api/cluster/status Retrying. round 2..!
E0708 04:12:29.402402   10456 avisession.go:860] CheckControllerStatus Error while generating http request https://localhost//api/cluster/status Get https://localhost//api/cluster/status: dial tcp 127.0.0.1:443: connect: connection refused
E0708 04:12:39.402646   10456 avisession.go:870] CheckControllerStatus Controller https://localhost//api/cluster/status Retrying. round 3..!
E0708 04:12:39.402756   10456 avisession.go:558] restRequest Error during checking controller state <nil>
E0708 04:12:39.402796   10456 avisession.go:582] Encountered client side error: &{Status:408 Request Timeout StatusCode:408 Proto: ProtoMajor:0 ProtoMinor:0 Header:map[] Body:<nil> ContentLength:0 TransferEncoding:[] Close:false Uncompressed:false Trailer:map[] Request:<nil> TLS:<nil>}
E0708 04:12:39.402974   10456 avisession.go:246] response error: Encountered an error on POST request to URL https://localhost/login: HTTP code: 408; error from Avi: 408 Request Timeout
--- FAIL: TestAviSessionControllerStatusCheckLimits (40.04s)
    **avisession_test.go:184: Session Creation failed: Encountered an error on POST request to URL https://localhost/login: HTTP code: 408; error from Avi: 408 Request Timeout**

_SDK not crashing now on unsuccessful controller status check._

**Pull request merged**